File: medeia-validator-java-examples/src/main/java/com/worldturner/medeia/examples/java/gson/objects/ReadObjectExample.java
package com.worldturner.medeia.examples.java.gson.objects;

import com.google.gson.Gson;
import com.google.gson.stream.JsonReader;
import com.worldturner.medeia.api.SchemaSource;
import com.worldturner.medeia.api.UrlSchemaSource;
import com.worldturner.medeia.api.ValidationFailedException;
import com.worldturner.medeia.api.gson.MedeiaGsonApi;
import com.worldturner.medeia.examples.java.domain.Person;
import com.worldturner.medeia.schema.validation.SchemaValidator;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.nio.charset.StandardCharsets;

public class ReadObjectExample {
    private static MedeiaGsonApi api = new MedeiaGsonApi();
    private static Gson gson = new Gson();

    private void parseValidExample() throws IOException {
        SchemaValidator validator = loadSchema();
        Reader dataReader =
                new InputStreamReader(
                        getClass().getResourceAsStream("/readobject/valid-person.json"),
                        StandardCharsets.UTF_8);
        JsonReader validatedReader = api.createJsonReader(validator, dataReader);
        Person person = gson.fromJson(validatedReader, Person.class);
        System.out.println(person.getFirstName());
    }

    private void parseInvalidExample() throws IOException {
        SchemaValidator validator = loadSchema();
        Reader dataReader =
                new InputStreamReader(
                        getClass().getResourceAsStream("/readobject/invalid-person.json"),
                        StandardCharsets.UTF_8);
        JsonReader validatedReader = api.createJsonReader(validator, dataReader);
        try {
            Person person = gson.fromJson(validatedReader, Person.class);
            throw new IllegalStateException("Invalid json data passed validation");
        } catch (ValidationFailedException e) {
            // Expected
            System.out.println("Validation failed as expected: " + e);
        }
    }

    @NotNull
    private SchemaValidator loadSchema() {
        SchemaSource source = new UrlSchemaSource(
                getClass().getResource("/readobject/person-address-schema.json"));
        return api.loadSchema(source);
    }

    public static void main(String[] args) throws IOException {
        ReadObjectExample example = new ReadObjectExample();
        example.parseValidExample();
        example.parseInvalidExample();
    }
}

File: medeia-validator-kotlin-examples/src/main/kotlin/com/worldturner/medeia/examples/kotlin/gson/objects/ReadObjectExample.kt
package com.worldturner.medeia.examples.kotlin.gson.objects

import com.google.gson.Gson
import com.worldturner.medeia.api.UrlSchemaSource
import com.worldturner.medeia.api.ValidationFailedException
import com.worldturner.medeia.api.gson.MedeiaGsonApi
import com.worldturner.medeia.examples.kotlin.domain.Person
import com.worldturner.medeia.schema.validation.SchemaValidator
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets

class ReadObjectExample {
    private val api = MedeiaGsonApi()
    private val gson = Gson()

    fun parseValidExample() {
        val validator = loadSchema()
        val dataReader = InputStreamReader(
            javaClass.getResourceAsStream("/readobject/valid-person.json"),
            StandardCharsets.UTF_8
        )
        val validatedReader = api.createJsonReader(validator, dataReader)
        val person = gson.fromJson<Person>(validatedReader, Person::class.java)
        println(person.firstName)
    }

    fun parseInvalidExample() {
        val validator = loadSchema()
        val dataReader = InputStreamReader(
            javaClass.getResourceAsStream("/readobject/invalid-person.json"),
            StandardCharsets.UTF_8
        )
        val validatedReader = api.createJsonReader(validator, dataReader)
        try {
            gson.fromJson<Person>(validatedReader, Person::class.java)
            throw IllegalStateException("Invalid json data passed validation")
        } catch (e: ValidationFailedException) {
            // Expected
            println("Validation failed as expected: $e")
        }
    }

    private fun loadSchema(): SchemaValidator {
        val source = UrlSchemaSource(
            javaClass.getResource("/readobject/person-address-schema.json")
        )
        return api.loadSchema(source)
    }
}

fun main(args: Array<String>) {
    ReadObjectExample().apply {
        parseValidExample()
        parseInvalidExample()
    }
}